pkg/storegateway/indexheader/lazy_binary_reader.go

```diff
@@ -230,125 +268,170 @@ func (r *LazyBinaryReader) EagerLoad() {
 	}
 
-	r.reader = reader
 	elapsed := time.Since(startTime)
 
 	level.Debug(r.logger).Log("msg", "lazy loaded index-header file", "path", r.filepath, "elapsed", elapsed)
 	r.metrics.loadDuration.Observe(elapsed.Seconds())
 
-	return nil
+	return reader, nil
 }
 
-// unloadIfIdleSince closes underlying BinaryReader if the reader is idle since given time (as unix nano). If idleSince is 0,
-// the check on the last usage is skipped. Calling this function on a already unloaded reader is a no-op.
-func (r *LazyBinaryReader) unloadIfIdleSince(ts int64) error {
-	r.readerMx.Lock()
-	defer r.readerMx.Unlock()
-
-	// Nothing to do if already unloaded.
-	if r.reader == nil {
-		return nil
-	}
-
-	// Do not unloadIfIdleSince if not idle.
-	if ts > 0 && r.usedAt.Load() > ts {
-		return errNotIdle
+func (r *LazyBinaryReader) waitAndCloseReader(req readerRequest) {
+	resp := <-req.response
+	if resp.reader != nil {
+		resp.inUse.Done()
 	}
+}
 
-	// Wait until all users finished using current reader.
-	r.readerInUse.Wait()
-
-	r.metrics.unloadCount.Inc()
-	if err := r.reader.Close(); err != nil {
-		r.metrics.unloadFailedCount.Inc()
-		return err
+// unloadIfIdleSince closes underlying BinaryReader if the reader is idle since given time (as unix nano). If idleSince is 0,
+// the check on the last usage is skipped. Calling this function on a already unloaded reader is a no-op.
+func (r *LazyBinaryReader) unloadIfIdleSince(tsNano int64) error {
+	req := unloadRequest{
+		// The channel is unbuffered because we will read the response. It should be buffered if we can give up before reading from it
+		response:       make(chan error),
+		idleSinceNanos: tsNano,
 	}
-
-	r.reader = nil
-	return nil
+	r.unloadReqC <- req
+	return <-req.response
 }
 
-// isIdleSince returns true if the reader is idle since given time (as unix nano).
-func (r *LazyBinaryReader) isIdleSince(ts int64) bool {
-	if r.usedAt.Load() > ts {
-		return false
+func (r *LazyBinaryReader) controlLoop() {
+	var loaded loadedReader
+	for {
+		select {
+		case readerReq := <-r.loadedReaderC:
+			if loaded.reader != nil {
+				loaded.inUse.Add(1)
+				r.usedAt.Store(time.Now().UnixNano())
+			}
+			readerReq.response <- loaded
+
+		case loadReq := <-r.loadReqC:
+			if loaded.reader == nil && loaded.err == nil {
+				loaded = loadedReader{}
+				loaded.reader, loaded.err = r.loadReader()
+				if loaded.reader != nil {
+					loaded.inUse = &sync.WaitGroup{}
+					r.usedAt.Store(time.Now().UnixNano())
+				}
+			}
+			close(loadReq.done)
+
+		case unloadPromise := <-r.unloadReqC:
+			if loaded.reader == nil {
+				// Nothing to do if already unloaded.
+				unloadPromise.response <- nil
+				continue
+			}
+
+			// Do not unloadIfIdleSince if not idle.
+			if ts := unloadPromise.idleSinceNanos; ts > 0 && r.usedAt.Load() > ts {
+				unloadPromise.response <- errNotIdle
+				continue
+			}
+
+			// Wait until all users finished using current reader.
+			loaded.inUse.Wait()
```

Comment:
How about we repace the `sync.WaitGroup` with this:
```go
type ctxWaitGroup struct {
	sync.WaitGroup
	once sync.Once
	done chan struct{}
}

func (w *ctxWaitGroup) Wait(ctx context.Context) error {
	w.once.Do(func() {
		go func() {
			w.WaitGroup.Wait()
			close(w.done)
		}()
	})
	select {
	case <-w.done:
		return nil
	case <-ctx.Done():
		return ctx.Err()
	}
}
```

And here do:
```suggestion
			if err := loaded.inUse.Wait(ctx); err != nil {
				unloadPromise.response <- err
			}
```

And we pass a context in the unload request ensuring that there's always a finite timeout (like 10 minutes, etc.)

Reply:
That's a nice solution but it combines a few different types of concurrency primitives and I'd like to see if we can simplify it a bit. It seems weird that there's nothing that does this (waitgroup + context) out of the box.  I don't think we need to block this PR for that so either way is fine to me: log or make a `WaitGroup` wrapper that respects contexts.

Reply:
so because this happens out of band there isn't a request context (well it's the process startup context, but that's not of a huge use). So I just hardcoded a 30 seconds wait for this. It will be logged here if it times out https://github.com/grafana/mimir/blob/63b5370dbea158847a0c90fa70f1d1ae24d53c57/pkg/storegateway/indexheader/reader_pool.go#L257


Does this sound ok?

Reply:
i realized that timeout is wrong :sob:

if we give up early there will be a goroutine which is waiting on the waitGroup.Wait. If after that comes a request which needs a reader, we will call WaitGroup.Add(1). It is illegal to call Add while Wait is still in progress and WaitGroup will panic.